storage_service/locations/models/dataverse.py

```diff
@@ -20,52 +22,120 @@
 
 
 class Dataverse(models.Model):
-    space = models.OneToOneField('Space', to_field='uuid')
-    host = models.CharField(max_length=256,
-        verbose_name=_l('Host'),
-        help_text=_l('Hostname of the Dataverse instance. Eg. apitest.dataverse.org'))
-    api_key = models.CharField(max_length=50,
-        verbose_name=_l('API key'),
-        help_text=_l('API key for Dataverse instance. Eg. b84d6b87-7b1e-4a30-a374-87191dbbbe2d'))
-    agent_name = models.CharField(max_length=50,
-        verbose_name=_l('Agent name'),
-        help_text=_l('Agent name for premis:agentName in Archivematica'))
-    agent_type = models.CharField(max_length=50,
-        verbose_name=_l('Agent type'),
-        help_text=_l('Agent type for premis:agentType in Archivematica'))
-    agent_identifier = models.CharField(max_length=256,
-        verbose_name=_l('Agent identifier'),
-        help_text=_l('URI agent identifier for premis:agentIdentifierValue in Archivematica'))
+    space = models.OneToOneField("Space", to_field="uuid")
+
+    host = models.CharField(
+        max_length=256,
+        verbose_name=_l("Host"),
+        help_text=_l(
+            "Hostname of the Dataverse instance. Eg. apitest.dataverse.org"
+        ),
+    )
+    api_key = models.CharField(
+        max_length=50,
+        verbose_name=_l("API key"),
+        help_text=_l(
+            "API key for Dataverse instance. Eg. "
+            "b84d6b87-7b1e-4a30-a374-87191dbbbe2d"
+        ),
+    )
+    agent_name = models.CharField(
+        max_length=50,
+        verbose_name=_l("Agent name"),
+        help_text=_l("Agent name for premis:agentName in Archivematica"),
+    )
+    agent_type = models.CharField(
+        max_length=50,
+        verbose_name=_l("Agent type"),
+        help_text=_l("Agent type for premis:agentType in Archivematica"),
+    )
+    agent_identifier = models.CharField(
+        max_length=256,
+        verbose_name=_l("Agent identifier"),
+        help_text=_l(
+            "URI agent identifier for premis:agentIdentifierValue "
+            "in Archivematica"
+        ),
+    )
     # FIXME disallow string in space.path
 
     class Meta:
         verbose_name = _l("Dataverse")
-        app_label = 'locations'
+        app_label = "locations"
+
+    ALLOWED_LOCATION_PURPOSE = [Location.TRANSFER_SOURCE]
+
+    def get_query_value(self, key, path, default=None):
+        """Retrieve a value from a string and then remove it where it exists so
+        that we can recurse through it for all the values we need.
+        """
```

Comment:
Hmm, the function just doesn't seem to work at all as anticipated. I will have a think about what a good fix would look like. I'd like to take as much work away from the user here as possible.

Reply:
This was a good catch. I've added it to the unit tests and then made a code change to fix it. I'd like to give the user as little as possible to do with configuring this and we've found some good flexibility with the current approach. Let me know what you think of the change, hopefully it is a little better. 

Ultimately, I'd like to see a custom form written for Dataverse where the values can be grabbed from the form elements instead. This was a little bit out of scope for the work done for OCUL this time around. Hopefully there is scope to make this a more robust solution in the near future.

Reply:
Yeah, this is probably fine. It just makes me uneasy because it's still brittle. I guess I don't have the implicit knowledge of what a Dataverse `dataset`, `subtree`, or `query_string` value can look like, but again it's not hard to imagine an input that breaks expectations here. Compare the happy:

    query: drinks_1 subtree: archivematica_1/1
    ('1', 'archivematica_1', 'drinks_1')

to the sad:

    query: drinks/1 subtree: archivematica_1/1
    ('1 subtree: archivematica_1/1', None, 'drinks')

Assuming the above is too edge-case to worry about at this stage, one thing you might still want to do is avoid the new kwarg and treat all keys the same by doing something like the following:

    import re

    def get_query_value(key, path, default=None):
        path = path.lower()
        value = path.split(key.lower(), 1)
        try:
            value = value[1].strip()
        except IndexError:
            value = default
        if value == "":
            value = default
        if value is None:
            path = path.replace(key, "")
        else:
            path = re.sub(key + "\s*" + value, "", path)
        return value, path.strip()

As implied by the above, `get_query_value` could be a `staticmethod` or a function cuz it doesn't need `self`.